### ISSUE TITLE
overloading numerals when algebra is open

### ISSUE THREAD

**Issue by avigad** [leanprover/lean3#809]
These work as expected:

``` lean
import data.finset data.nat data.int algebra.ring
open finset nat int

check '{1, 2, 3}               -- finset ℕ
check ('{1, 2, 3} : finset ℕ)  -- finset ℕ
check ('{1, 2, 3} : finset ℤ)  -- finset ℤ

check insert 1 (insert 2 (insert 3 empty)) -- finset ℕ

definition foo_nat (x : finset ℕ) : finset ℕ := x
definition foo_int (x : finset ℤ) : finset ℤ := x

check foo_nat '{1, 2, 3}  -- finset ℕ
check foo_int '{1, 2, 3}  -- finset ℤ

```

But if `algebra` is added anywhere to the `open` command, all the checks fail. It looks like the elaborator no longer considers coercions.

**Comment by dselsam** [comment]
The following simple example fails as well:

``` Lean
check @finset.insert _ _ 1 (@finset.empty ℕ)
```

The issue is that the unifier attempts to disambiguate the `1` before it attempts to unify `@finset ?M ~ @finset ℕ`, and when it does it does not consider coercions on the choices for `1` since the target type is not known yet:

https://github.com/leanprover/lean/blob/master/src/frontends/lean/elaborator.cpp#L111

This restriction is presumably for performance reasons, though I don't know how important it is.

The interaction between choice case-splits, delta case-splits, and coercions can be subtle, but in this case there is a simple solution: we can set `finset` to be irreducible (after proving basic properties about it) so that `@finset ?M ~ @finset ℕ` unifies immediately. I suspect that this will be a common pattern, and so we might consider a new keyword or attribute that means "irreducible outside of the namespace".

**Comment by avigad** [comment]
Cool. Adding `attribute finset [irreducible]` solves the problem for all the examples but these:

``` lean
check '{1, 2, 3}
check insert 1 (insert 2 (insert 3 empty))
```

In my book, that's perfectly reasonable. Let's let Leo weigh in before closing the issue, but I am happy with this solution.

**Comment by dselsam** [comment]
Whoops, I did not notice that those examples still did not work. Removing the meta-variable check fixes them all, and doesn't seem to noticeably affect performance (though I did not profile carefully). I see no reason not to remove the check and support the general case.

Note: removing the check breaks one line in one proof:

https://github.com/dselsam/lean/blob/master/library/theories/number_theory/irrational_roots.lean#L90,

but it is easily fixed by changing the second argument to `rewrite` from `'a = 0'` to `this`.

**Comment by leodemoura** [comment]
@dselsam Thanks for looking at this issue.
I used an alternative workaround described in the commit above.

I agree the interaction between choice case-splits, delta case-splits, and coercions is subtle. The behavior is quite brittle sometimes. I'm sure we have other instances of the problem described in this issue. Suggestions and improvements are very welcome.

**Comment by dselsam** [comment]
> I believe the following condition
> https://github.com/leanprover/lean/blob/master/src/frontends/lean/elaborator.cpp#L111
> reduces counter-intuitive behavior. Example, the coercion should not
> influence the resulting type.

Do you mind expanding on this? My intuition is that if there is no solution otherwise, it would be appropriate to "salvage" elaboration using the coercion. On the other hand, if backtracking further would yield a "better" solution, I understand that it would be problematic to let the questionable coercion hijack elaboration. I am not suggesting this, but might it be semantically acceptable to put these coercions at the bottom of the case-split stack instead of the top?

> The problem is that many files in the standard library are abusing the higher-order unification procedure

Do you mind explaining what constitutes "abusing" higher-order unification?

**Comment by leodemoura** [comment]
> Do you mind expanding on this? 

Suppose we have a term `t` and it should have type `?M a b c` where `?M` is a metavariable.
In the current implementation, we will not inject a coercion until `?M` is assigned.
If it cannot be inferred before we start processing "expensive constraints" (e.g., flex-rigid constraints), we don't consider coercions for `t`.
Now, suppose we change the implementation and inject a coercion `c` even if `?M a b c` is not yet known. Then the resulting term would be `c t` and we may be producing a term with a type different from the one intended by the user. 

> On the other hand, if backtracking further would yield a "better" solution

We don't produce multiple solutions. We stop at the first solution we find.
I think it is currently not feasible to produce many solutions and rank them.
Coercions are problematic because they produce a lot of ambiguity.

> I am not suggesting this, but might it be semantically acceptable to put these coercions at the bottom of the case-split stack instead of the top?

This is not feasible for performance reasons. Note that we currently try to solve coercions before we start to process expensive constraints such as flex-rigid ones. 

> Do you mind explaining what constitutes "abusing" higher-order unification?

I say users are "abusing" the higher-order unification procedure when they rely too much on definitional equality (delta-reduction in particular).
Higher-order unification is an undecidable problem even when we don't consider delta and iota-reduction.
Here is an example I rewrote yesterday.

``` lean
theorem of_nat_sub {m n : ℕ} (H : m ≥ n) : m - n = sub m n :=
(sub_eq_of_eq_add (!congr_arg (nat.sub_add_cancel H)⁻¹))⁻¹
```

and here is the new expanded one.

``` lean
theorem of_nat_sub {m n : ℕ} (H : m ≥ n) : m - n = sub m n :=
assert m - n + n = m,     from nat.sub_add_cancel H,
begin
  symmetry,
  apply sub_eq_of_eq_add,
  rewrite [-of_nat_add, this]
end
```

The old one depends on the definitions of `int.add`, `nat.add` and `of_nat`, and it broke when I changed the unifier yesterday. To make the proof more robust I explicitly added the rewrite `of_nat_add`. 
The new proof doesn't depend on delta-reductions as much anymore.

**Comment by dselsam** [comment]
> We don't produce multiple solutions. We stop at the first solution we find.
> I think it is currently not feasible to produce many solutions and rank them.

I understand that, but it still seems that the concern of reaching a 'poor' solution does impact the elaborator, since the reason you do not e.g. inject a coercion until `?M` is assigned is because of the concern that it might shadow a better solution that might be obtained by backtracking. If there is no other solution, then considering the coercion does not introduce any ambiguity (though we may not know if there are other solutions when we need to make the decision, and we do not want to take the risk).

Am I still misunderstanding something? 

> Suppose we have a term t and it should have type ?M a b c where ?M is a metavariable.
> In the current implementation, we will not inject a coercion until ?M is assigned.

I don't think this is true. Consider the following code:

``` Lean
import data.finset data.nat
open finset nat 

check insert 1 (insert 2 (insert 3 empty)) -- finset ℕ
```

Without `algebra` open, this type checks as `finset ℕ`, even though as far as I can tell the only signal for `ℕ` comes from the coercion itself. I think it is only when the term in question is part of a choice constraint and wakes up to find a meta target that the preemptive coercions are not considered. 

This behavior seems inconsistent though, because the coercion might still shadow a better solution that could have been achieved by backtracking to an earlier choice point, even if the target type is known when the choice point in question wakes up:

``` Lean
import data.finset data.nat
open finset nat 

notation `***` := ℕ
notation `***` := bool

notation `+++` := bool.ff
notation `+++` := num.zero

set_option pp.notation true
check @insert *** _ +++ empty -- finset ℕ
```

Zooming out, here is the mental model of the elaborator that I would like to have:
1. The user enters a (pre)-expression: `t1 (t2 t3)`
2. Each of the atoms `t1`,`t2`,`t3` might refer to one of several different expressions due to ad-hoc overloading and coercions, and these sets are fixed and known by the user: `{x1,x2} ({x3} {x5,x6,x7})`.
3. For each complete instantiation to the these atoms, there will be some additional type-dependent `_` placeholders that need to be inserted.
4. The elaborator will pick one element from each of these sets, and instantiate the corresponding `_`s with arbitrary expressions, so that the resulting term type checks.
5. If it fails to do so even though there is a satisfying assignment, it is almost always because the elaborator did not unfold enough, and you can debug it by finding steps that required hidden unfoldings and either making the necessary definitions reducible or by adding intermediate steps.

Zooming out further, I think that supporting all of this ambiguity is a fascinating challenge, but that in the long run the convenience of the ambiguity may not be worth the complexity. I may oscillate on this and am not suggesting it, but for most use cases I think I might even prefer to give up the possibility of ambiguity for the "relevant" terms entirely (i.e. get rid of ad-hoc overloading and coercions), if it meant we could have a simple and uniform elaborator (although I haven't given up hope that we can achieve this simplicity while still supporting all of these features). Leo made a good point in #800 that a commitment to LaTeX-quality rendering may more than compensate for clumsy input syntax.

It may also be possible to split the elaborator into two: the core elaborator that only instantiates the `_`s and that is always used, and then an optional layer above that that deals with user-level ambiguity and calls down to the core elaborator as a gray box.

(Please note: removing user-level ambiguity would not entail that users had to write e.g. `nat.of_num 1`, since specific coercions could be registered as 'always apply' within a namespace without introducing ambiguity.)

**Comment by avigad** [comment]
A general comment: the elaborator is already a very complex piece of code, and Leo has been tinkering with it for more than two years now. I worry that if we keep fixating on it, we are letting the perfect be the enemy of the good. It may be the case that improvements are possible, but it is not clear to me that spending a lot of time on it will increase usability much, and there are things that are much more important, like automation and the simplifier.

Coercions are particularly vexing, because expressions that rely on them are often simply ambiguous, with no obvious "correct" reading. In that case, I think we need to find usage patterns that reduce the ambiguity, rather than training the elaborator to read our minds.

Would it make sense to put off this kind of experimentation for a rainy day, when automation and other core parts of the system are already in place? At worst, it will require users to make their expressions more explicit in the interim. (In the future, we may be induced to make changes to the elaborator that break things in the library, but those instances should be minimal; if such a change fails to check expressions that were checkable before, that is evidence that the change has introduced counterintuitive behavior that was not there before.)

**Comment by leodemoura** [comment]
> I don't think this is true. Consider the following code:

I should have been more precise. We consider two types when we are injecting a coercion: the type of the term `t`, and the expected type where `t` occurs.
We only consider a coercion when the type of `t` is known (i.e., it is _not_ of the form `?M ...`).
In your example, the type of the term `t` (i.e., `1`, `2` and `3` in the example) is known (it is a `num`). 
Here is the code I was referring to:
https://github.com/leanprover/lean/blob/master/src/frontends/lean/coercion_elaborator.cpp#L102-L118

> Zooming out, here is the mental model of the elaborator that I would like to have:

I don't see how to implement step 2. To inject the implicit arguments we need to know the types of `t1` and `t2`. To know the type of `t1` and `t2` we need to resolve the choices.
For example, if the `algebra` and `nat` namespaces are open, then `a + b` has two possible interpretations:
1- `has_add.add _ _ a b`
2- `nat.add a b`

We have the same problem with coercions. The number of implicit arguments depends on the coercion selected.

> I think I might even prefer to give up the possibility of ambiguity for the "relevant" terms entirely (i.e. get rid of ad-hoc overloading and coercions)

The problem is that both features are extensively used. 

We say a term `t` is "relevant" if the type of `t` is **not** a subsingleton.
So, we only really know whether a term is "relevant" or not after we elaborate it.

> if it meant we could have a simple and uniform elaborator (although I haven't given up hope that we can achieve this simplicity while still supporting all of these features).

It would be great to have a simple and uniform elaborator, but it would have to support most of the features we currently provide. Several of the current features were based on actual user feedback/request. It would be bad to remove them.

> It may also be possible to split the elaborator into two: the core elaborator that only instantiates the _s and that is always used, and then an optional layer above that that deals with user-level ambiguity and calls down to the core elaborator as a gray box.

The elaborator is composed of different modules. However, these modules are **not** executed in sequence. I agree it would be nice to have this kind of architecture, but I don't see how to do it and support the features we want. For example, the `a+b` example above demonstrates the interaction between adding implicit arguments, inferring types and resolving choices (coercions and ad-hoc polymorhism).

**Comment by leodemoura** [comment]
> it is not clear to me that spending a lot of time on it will increase usability much, and there are things that are much more important, like automation and the simplifier.

I agree. 

I think the only motivation for investing more time on the elaborator is stability and predictability. 
This will be important as the user base grows.

> Would it make sense to put off this kind of experimentation for a rainy day, when automation and other core parts of the system are already in place? 

I agree.

BTW, I'm also burnout. I don't really have energy to invest on the elaborator :-(
I want to work on automation and code generation for Lean.

@dselsam I'm not saying we shouldn't invest time on the elaborator, nor I'm implying this is not an important problem. If you are interested in working on this problem, I would be happy to help you and explain the problems I had, but you will have to be one "pushing it forward".

**Comment by dselsam** [comment]
@leodemoura 

> I don't see how to implement step 2.

Good point. That was an error on my part, and I updated the mental model accordingly.

> BTW, I'm also burnout. I don't really have energy to invest on the elaborator :-(

I understand, and I am not trying to add things to your plate! I appreciate the thoughtful responses very much, but as a rule, anything I feel strongly about after discussing I would be happy to implement myself. 

Also, I suspect that the complexity of the elaborator contributes to the burden of maintaining the system. Part of my sudden interest in simplifying the elaborator is because I recently hit my own complexity limit where the mental effort required to understand what the elaborator was doing became too draining.

The elaborator also has a large surface area in the system, and even the automation procedures will need to interact with it, which might cause additional complications down the line.

@avigad 

> it is not clear to me that spending a lot of time on it will increase usability much, and there are things that are much more important, like automation and the simplifier.

I understand your perspective, but I’ll just point out that you are a super-power-user of Lean and helped design it. As far as I know, almost everybody who works in Lean now has either helped design it or works closely with someone who did. I think you may be underestimating how hard it might be for even dedicated newcomers to master.

> Would it make sense to put off this kind of experimentation for a rainy day, when automation and other core parts of the system are already in place?

Maybe, but I tend to think we are at the tail end of the "constitutional" phase of the system, and that it will only get harder and harder to change the foundations.

I am also not particularly worried about breaking the standard library, since we still have the fully elaborated terms and in the worst case could just write a simple un-elaborator to replace any broken proofs.

---

Thanks to your feedback and clarifications, I believe I understand the issues and trade-offs involved now. I will brew on it.

**Comment by leodemoura** [comment]
> I suspect that the complexity of the elaborator contributes to the burden of maintaining the system. 

You are correct. 
Moreover, we are learning the "requirements" (e.g., necessary features, issues, etc) as we go. 

> I recently hit my own complexity limit where the mental effort required to understand what the elaborator was doing became too draining.

Note that most users will not try to understand, fix and modify the source code like you do.
My point is that you have "two hats": user and developer.
Some of the complexity is because you use the developer hat too.

That being said, I think it would be nice to (try to) reduce the complexity.
For example we can reduce the complexity by cleaning up the `library/unifier.cpp` module.
This file have been "evolving" for quite some time. It can be simplified.
Another place to simplify is `frontends/lean/elaborator.cpp`. 
The other complicated module is the one that compiles recursive equations into recursors.
The rest of the code base is simpler and more modular.
For example, the tactic framework is very modular and the tactics are completely independent of each other.

Another approach is to design a new simpler and more powerful elaboration procedure that is easy to maintain and understand. 

> The elaborator also has a large surface area in the system, and even the automation procedures will need to interact with it, which might cause additional complications down the line.

Automation like `blast` should not interact with the elaborator. 
It gets a fully elaborated context/goal, and has to synthesize a term of the expected type.
We can also implement a simpler unifier/matcher to be used in proof search procedures.
The requirements are different. The current unifier has a lot of complexity because it has to interact with several modules (e.g., coercion resolution and adhoc polymorphism via choice constraints).
We can avoid that in the automatic proof search procedures. 

> I think you may be underestimating how hard it might be for even dedicated newcomers to master.

Some remarks:
- It is much harder to design a reusable Lean library than using it (e.g., it is harder to design the algebraic hierarchy than using it).
- A newcomer doesn't have to master everything to be able to use Lean. This is the case for complex programming languages like C++. For example, I don't really know all coercion resolution rules used in C++. 
- We can add configuration options that disable complex features.

> Maybe, but I tend to think we are at the tail end of the "constitutional" phase of the system, and that it will only get harder and harder to change the foundations.

I agree. It will get harder and harder to change the elaborator as the user base grows.

> we still have the fully elaborated terms and in the worst case could just write a simple un-elaborator to replace any broken proofs.

Yes, this is a must-have feature. Floris and I talked about having a "conservative" pretty printer.
This kind of pretty printer would avoid "overloaded" notation, include any implicit argument that requires higher order unification to be inferred, etc.
The idea is to output a term that can be elaborated without performing search, and it is not too big.

**Comment by avigad** [comment]
Don't get me wrong; I realize that when the elaborator fails it can be frustrating for users when they don't know why it isn't working. It is just that I doubt there is a perfect solution, and I think what we have is pretty close to optimal. I would be happy to be proved wrong, though.

In a less-than perfect world, I will settle for an elaborator that works most of the time, if the failures are relatively easy to diagnose and fix by adding a little more information. (It is not so bad if users don't know exactly why it isn't working, as long as they can figure out what to do to fix it.)

I will admit that the issues with the overloaded 0 and 1 in algebra are vexing. Would it help to put them in a namespace (like algebra.num_notation) that users would have to open explicitly? (There isn't any way to open algebra with all the notation _except_ 0 and 1, is there?)
